### PR TITLE
Refine the borsh implementation

### PR DESCRIPTION
**Pull request by danlehmann**
This removes the vec! as we don't want to allocate.

Also add a lot of tests to ensure various corner cases are covered

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1087,31 +1081,35 @@ where
...
 where
     Self: Number,
 {
     fn deserialize_reader<R: borsh::io::Read>(reader: &mut R) -> borsh::io::Result<Self> {
-        let mut buf = vec![0u8; core::mem::size_of::<T>()];
+        // Ideally, we'd want a buffer of size `BITS >> 3` or `size_of::<T>`, but that's not possible
+        // with arrays.
+        // So instead we'll do a 16 byte buffer which handles the largest arbitrary-ints possible -
+        // not ideal, but still pretty small and better than going through an allocator.
+        let mut buf = [0u8; 16];
```
Comment (danlehmann) [inline src/lib.rs:1112]: hmm this might be wrong - we might be asking too much from Borsh here. Will add a test to catch this and fix the implementation

**Review by dzmitry-lahoda — APPROVED** [review]
